### PR TITLE
[Validator] Add the When constraint and validator docs

### PR DESCRIPTION
**Pull request by wuchen90**

### CODE REVIEW

File: reference/constraints/When.rst [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+properties::
+
+    // src/Model/Discount.php
+    namespace App\Model;
+
+    use Symfony\Component\Validator\Constraints as Assert;
+
+    class Discount
+    {
+        private $type;
```
Comment (94noni) [inline reference/constraints/When.rst]: you can leverage type hinting here and below (for better doc as sf6+ requires php8+)

File: reference/constraints/When.rst [hunk 2]
```diff
@@ -0,0 +1,202 @@
+When
+===========
+
+This constraint allows you to apply constraints validation only if the provided expression returns true.
```
Comment (94noni) [inline reference/constraints/When.rst]: you should define a version added directive also
have a look at https://github.com/javiereguiluz/symfony-docs/commit/f32ee434c78f0210b1630aecf31b2c89a80f0074#diff-f552d75abf80f4985e72c4fe4d6664535dc95395a289456cf102f659bde0cfdfR4 as example

File: reference/constraints/When.rst [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Discount
+        {
+            /**
+             * @Assert\GreaterThan(0)
+             * @Assert\When(
+             *     expression="this.type == 'percent'",
+             *     constraints={@LessThan(100, message="The value should be between 0 and 100!")}
```
Comment (94noni) [inline reference/constraints/When.rst]: what about juste writing `The value should be less than 100`
as it is the `@Assert\GreaterThan(0)` that validates the 0

File: reference/constraints/When.rst [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+To learn more about the expression language syntax, see
+:doc:`/components/expression_language/syntax`.
+
+Inside of the expression, you have access to up to 2 variables:
+
+Depending on how you use the constraint, you have access to 1 or 2 variables
+in your expression:
+
+* ``this``: The object being validated (e.g. an instance of Discount);
+* ``value``: The value of the property being validated (only available when
```
Comment (94noni) [inline reference/constraints/When.rst]: could be great to give a code example using the `value` :)

File: reference/constraints/When.rst (hunk 3 above)
Comment (wuchen90) [reply reference/constraints/When.rst]: It's just an example and in my opinion, this message is more user friendly.

File: reference/constraints/When.rst (hunk 4 above)
Comment (wuchen90) [reply reference/constraints/When.rst]: Done.

File: reference/constraints/When.rst [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        namespace App\Model;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Discount
+        {
+            #[Assert\GreaterThan(0)]
+            #[Assert\When(
+                expression: 'this.type == "percent"',
+                constraints: [new Assert\LessThan(100, message: 'The value should be between 0 and 100!')],
```
Comment (xabbuh) [inline reference/constraints/When.rst]:
```suggestion
                constraints: [new Assert\LessThan(100, message: 'The value should be between 0 and 99!')],
```

According to the rest of the example. But maybe allowing 100 everywhere might make more sense.

Comment (wuchen90) [reply reference/constraints/When.rst]: Let's change to `LessThanOrEqual`.

**Comment by wouterj** [comment]
This is a cool feature! Thank you @wuchen90

**Pull request merged**